### ISSUE TITLE
minifier: mangle properties

### ISSUE THREAD

**Issue by sapphi-red** [oxc-project/oxc#15375]
Property names may take up the large part of the output size. Other minifiers have the mangle properties feature for those cases.

- esbuild: https://esbuild.github.io/api/#mangle-props
- SWC: https://swc.rs/docs/configuration/minification#jscminifymangleproperties
- terser: https://terser.org/docs/options/#mangle-properties-options
- uglifyjs: https://github.com/mishoo/UglifyJS/blob/master/README.md#cli-mangling-property-names---mangle-props

(I'm creating this issue as a placeholder, if you actually need this feature, make sure to add "👍" and describe the usecase in the comments)

**Comment by logaretm** [comment]
I'm doing an attempt at migrating to [rolldown in Sentry's JavaScript repo,](https://github.com/getsentry/sentry-javascript/pull/18156) it would be great if we can have fine grained properties mangling. We have been using `properties.regex` and `properties.reserved` options to prevent specific properties from getting mangled.

We also use the top-level `mangle.reserved` but I don't think we need it. None of that should be needed actually, but just from the POV of matching the output with the rollup setup.

**Comment by maxpatiiuk** [comment]
Authoring an internal framework (Lit-based). The runtime size is 13kb gzip (tiny, thanks to optimizations, and Lit itself being small)

I am using private fields for properties that don't need to be accessed outside the class. But, I do need some properties to be accessible by the runtime code outside the class, even if it is not a public API. I use property mangling for those.

<details>
<summary>My "esbuild" config in vite.config.ts (lib mode)</summary>

```tsx
  esbuild: {
    // eslint-disable-next-line require-unicode-regexp
    mangleProps: /^_/,
    /**
     * Vite runs ESBuild transform on one file at at time. Thus, ESBuild does
     * not know what was a given property name mangled to in a different file.
     * To avoid collisions, and ensure private property names stay stable
     * between builds, we use a mangle cache.
     */
    /* eslint-disable @typescript-eslint/naming-convention */
    mangleCache: {
      // Comment shows who defines the property (rather than who uses it).
      // This way we can prevent collisions when two properties are mangled to
      // the same name on the same object.

      // Making the property names more verbose to avoid accidental collisions
      // between different objects.

      // Because subclass cannot re-use a private variable name that is already
      // used by a parent class, we are using __ over _ in classes that are
      // commonly sub-classed (LitElement and Controller) to prevent reserving
      // property names that consumers may want to use. Similarly, if we
      // introduce a new private property, we can break consumers if the name
      // collides with what they are already using, so __ is useful.

      // ProxyComponent
      _LitElementConstructor: "A",
      _LitElementLoadPromise: "B",
      _proxiedProperties: "C",
      _proxiedSyncMethods: "D",
      _proxiedAsyncMethods: "E",
      _initializeProxyPrototype: "F",
      _ancestorLoadPromise: "G",
      _postLoadedDeferred: "H",

      // ProxyComponent and LitElement
      __postLoadDeferred: "I",
      __offspringComponents: "J",
      __runtime: "K",
      __tagName: "L",

      // LitElement
      _controllers: "M",
      __lazy: "N",

      // Controller
      __assignedProperty: "O",
      __ready: "P",

      // ControllerManager (extends Controller)
      _changedProperties: "Q",
      _isEnforcingReadonly: "R",
      _firePropTrackers: "S",
      _keyTrackers: "T",
      _setAutoDestroyTimeout: "U",
      _trackedValue: "V",
      _markExports: "W",
      _resolveExports: "X",

      // useAccessor
      _boundAccessorProperties: "Y",
      _createAccessorInstance: "Z",
      _currentlyBindingPropertyName: "A",

      // If you run out of alphabet, it is safe to reuse letters between
      // different objects.
    },
    /* eslint-enable @typescript-eslint/naming-convention */
  },
```

</details>

**Comment by DylanPiercey** [comment]
Just tried swapping out esbuild for building the [marko 6 runtime](https://marko-js.com).

Ran into two cases (this and #6073) where the current oxc produces larger bundles vs esbuild.

References:
[potential rolldown setup](https://github.com/marko-js/marko/blob/62664f8d9eac97ba56585c281727333674a1c7e6/packages/runtime-tags/scripts/bundle.mts)
[current esbuild setup](https://github.com/marko-js/marko/blob/62664f8d9eac97ba56585c281727333674a1c7e6/packages/runtime-tags/scripts/bundle.ts)

Would love to see this added ❤️

**Comment by spaceemotion** [comment]
Would love to has this feature for package-internal classes that are never exposed to the public API, for example: custom iterators, state objects that have helper methods, adapters that wrap things.

For example:
```ts
class PausableIterator<T> implements IterableIterator<T> {
  private done = false;

  public constructor(
    private readonly resume: () => boolean,
    private readonly getResult: () => T | undefined,
  ) {}

  public next(): IteratorResult<T> {
    if (this.done) return ITER_DONE;

    const finished = this.resume();
    const value = this.getResult();

    if (value === undefined) {
      this.done = true;
      return ITER_DONE;
    }

    if (finished) this.done = true;

    return { value, done: false };
  }

  public [Symbol.iterator](): IterableIterator<T> {
    return this;
  }
}
```
can probably be minified/mangled to:
```ts
class PausableIterator<T> implements IterableIterator<T> {
  private a = false;

  public constructor(
    private readonly b: () => boolean,
    private readonly c: () => T | undefined,
  ) {}

  public next(): IteratorResult<T> {
    if (this.a) return ITER_DONE;

    const d = this.b();
    const e = this.c();

    if (e === undefined) {
      this.a = true;
      return ITER_DONE;
    }

    if (d) this.a = true;

    return { value, done: false };
  }

  public [Symbol.iterator](): IterableIterator<T> {
    return this;
  }
}
```

Not sure how to best handle the mangling of these as the flow analysis of whether an instance of it gets actually exposed might be hard?

**Comment by Boshen** [comment]
Implemented by #24740 and exposed through `oxc-minify` by #24741. Released in v0.145.0 via #25865.

Downstream Rolldown integration is tracked separately in rolldown/rolldown#10374.

Closing as completed.
